### DISCUSSION TITLE
Python UDF poor cpu usage

### DISCUSSION THREAD

**Discussion by caiocesarrm** [duckdb/duckdb#8206] [Q&A] [answered]
Hi everyone!

i have these two python UDF's:

```python
def b64_uuid5(namespace: str, name: str) -> str:
    namespace = uuid.uuid5(uuid.NAMESPACE_OID, namespace)
    return base64.urlsafe_b64encode(uuid.uuid5(namespace, name).bytes_le).decode("utf-8")

def normalize(value: str) -> str:
    if value:
        return value.strip(". ").lower()
```

usage example:

```python
con = duckdb.connect(database=":memory:", read_only=False)

con.execute("""
SET memory_limit='50GB';
SET threads TO 24;
SET external_threads TO 24;
SET enable_progress_bar=true;
""")

con.create_function('b64_uuid5', b64_uuid5, [VARCHAR, VARCHAR], VARCHAR, side_effects=False)
con.create_function('normalize', normalize, [VARCHAR], VARCHAR, side_effects=False)

con.execute(f"""
COPY 
(
    select
        b64_uuid5('namespace', column_name) as uuid
    FROM read_parquet('{folder}/data/*.parquet')
)
TO '{folder}/graph/tmp.parquet' (FORMAT 'parquet');
""")
```

Tried without setting any config, with and without in memory db. The query takes forever, tried with pyspark UDF and 300 million rows finish in 2~3 minutes.

CPU usage is pretty low across all cores.

DuckDB CPU usage
![duck](https://github.com/duckdb/duckdb/assets/41218670/79f5f43c-18e9-4421-81c7-14b986a56aaf)

Spark CPU usage
![spark](https://github.com/duckdb/duckdb/assets/41218670/28d58374-a9ea-4361-9d36-8aaa010c7b0b)

Python 3.8
duckdb 0.8.1

I'm doing something wrong? I'm not really interested in solving the specific problem, like using some native uuid5 or native text manipulation. My concern is more generally about the state of the Python UDF.

Thanks in advance!

**Comment by Tishj** [comment, marked as answer]
The native python UDFs are not parallelized very well yet, as we need to do a function call for every tuple, and the GIL needs to be held while doing that.

We hope to alleviate this issue in the future, but in the meantime we recommend using the pyarrow variant of our UDFs, see the docs for more information on this

**Reply by wlxiong** [reply]
I tested the benchmark in the official [blog](https://duckdb.org/2023/07/07/python-udf.html#built-in-python-vs-pyarrow) in [this colab notebook](https://colab.research.google.com/drive/1HRIca7pkuZ9exbwoqNsuaShkclJzUYUt?usp=sharing). No significant improvement is observed. Any idea?

```
Built-In:
Time: 4.460603789000743
[(50000005000000,)]
PyArrow:
Time: 3.654671612999664
[(50000005000000,)]
```

On my devbox running Linux on AMD CPU, I found that pyarrow UDFs only utilize 1-2 CPU cores. It seems like the pyarrow UDFs do not run in parallel.

**Reply by Tishj** [reply]
Hmm that feels like a regression?
But also we haven't really looked into optimizing the performance of UDFs yet, the idea behind the UDFs is that it can fill in the gaps functionality-wise, but there is no promise of it being performant.

Though we will likely look to optimize it better in the future, we have plans to make use of the GIL-per-subinterpreter feature added in python12 that will hopefully allow us to parallelize much better

**Comment by wlxiong** [comment]
The announcement blog post said it was tested on Apple M1 CPU. One of my colleague happened to have one. He finds pyarrow UDF is faster than native UDF on M1. It’s interesting that the benchmark runs faster on M1 than on AMD server grade CPU. Maybe M1 has higher effective memory throughput.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
